libraries/chain/db_market.cpp

```diff
@@ -191,27 +191,101 @@ void database::cancel_order(const force_settlement_object& order, bool create_vi
    remove(order);
 }
 
-void database::cancel_order( const limit_order_object& order, bool create_virtual_op  )
+void database::cancel_limit_order( const limit_order_object& order, bool create_virtual_op, bool skip_cancel_fee )
 {
-   auto refunded = order.amount_for_sale();
-
-   modify( order.seller(*this).statistics(*this),[&]( account_statistics_object& obj ){
-      if( refunded.asset_id == asset_id_type() )
+   // if need to create a virtual op, try deduct a cancellation fee here.
+   // there are two scenarios when order is cancelled and need to create a virtual op:
+   // 1. due to expiration: always deduct a fee if there is any fee deferred
+   // 2. due to cull_small: deduct a fee after hard fork 604, but not before (will set skip_cancel_fee)
+   const account_statistics_object* seller_acc_stats = nullptr;
+   const asset_dynamic_data_object* fee_asset_dyn_data = nullptr;
+   limit_order_cancel_operation vop;
+   share_type deferred_fee = order.deferred_fee;
+   asset deferred_paid_fee = order.deferred_paid_fee;
+   if( create_virtual_op )
+   {
+      vop.order = order.id;
+      vop.fee_paying_account = order.seller;
+      // only deduct fee if not skipping fee, and there is any fee deferred
+      // TODO check if skip_fee is enabled?
```

Comment:
Please resolve this TODO.

Reply:
I found nowhere in the project has set `skip_fee` to `true`, so not sure what it is being used for. Even if it's enabled, in `evaluator::apply()`, the fee will be deducted from the payer.

Also, `skip_fee` is a member of `transaction_evaluation_state`, I don't know when we need to use it, actually in this PR I changed `database::clear_expired_orders()` from using `with_skip_flags(...)` 
 (with a `transaction_evaluation_state` object) to simply call `cancel_limit_orders()`, which looks simpler for me.

Thoughts?